identity/api_test.go

```diff
@@ -194,36 +195,81 @@ func TestCrashAfterRevocation(t *testing.T) {
 	c1 := NewIdentity(el, 2, "one")
 	c2 := NewIdentity(el, 2, "two")
 	c3 := NewIdentity(el, 2, "three")
-	defer c1.Close()
-	defer c2.Close()
-	defer c3.Close()
+	defer c1.Client.Close()
+	defer c2.Client.Close()
+	defer c3.Client.Close()
 	log.ErrFatal(c1.CreateIdentity())
 	log.ErrFatal(c2.AttachToIdentity(c1.ID))
 	proposeUpVote(c1)
 	log.ErrFatal(c3.AttachToIdentity(c1.ID))
 	proposeUpVote(c1)
 	proposeUpVote(c2)
-	log.ErrFatal(c1.ConfigUpdate())
-	log.Lvl2(c1.Config)
+	log.ErrFatal(c1.DataUpdate())
+	log.Lvl2(c1.Data)
 
-	conf := c1.GetProposed()
-	delete(conf.Device, "three")
-	log.Lvl2(conf)
-	log.ErrFatal(c1.ProposeSend(conf))
+	data := c1.GetProposed()
+	delete(data.Device, "three")
+	log.Lvl2(data)
+	log.ErrFatal(c1.ProposeSend(data))
 	proposeUpVote(c1)
 	proposeUpVote(c2)
-	log.ErrFatal(c1.ConfigUpdate())
-	log.Lvl2(c1.Config)
+	log.ErrFatal(c1.DataUpdate())
+	log.Lvl2(c1.Data)
 
 	log.Lvl1("C3 trying to send anyway")
-	conf = c3.GetProposed()
-	c3.ProposeSend(conf)
+	data = c3.GetProposed()
+	c3.ProposeSend(data)
 	if c3.ProposeVote(true) == nil {
 		t.Fatal("Should not be able to vote")
 	}
 	log.ErrFatal(c1.ProposeUpdate())
 }
 
+func TestVerificationFunction(t *testing.T) {
+	l := onet.NewTCPTest()
+	hosts, el, _ := l.GenTree(2, true)
+	services := l.GetServices(hosts, identityService)
+	s0 := services[0].(*Service)
+	defer l.CloseAll()
+	for _, s := range services {
+		log.Lvl3(s.(*Service).Identities)
+	}
+
+	c1 := NewTestIdentity(el, 50, "one1", l)
+	log.ErrFatal(c1.CreateIdentity())
+
+	// Hack: create own data-structure with twice our signature
+	// and send it directly to the skipblock. Without a proper
+	// verification-function, this should pass.
+	data2 := c1.Data.Copy()
+	kp2 := config.NewKeyPair(network.Suite)
+	data2.Device["two2"] = &Device{kp2.Public}
+	data2.Storage["two2"] = "public2"
+	hash, err := data2.Hash()
+	log.ErrFatal(err)
+	sig, err := crypto.SignSchnorr(network.Suite, kp2.Secret, hash)
+	log.ErrFatal(err)
+	data2.Votes["one1"] = &sig
+	data2.Votes["two2"] = &sig
+	id := s0.getIdentityStorage(c1.ID)
+	require.NotNil(t, id, "Didn't find identity")
+	_, cerr := s0.skipchain.StoreSkipBlock(id.SCData, nil, data2)
+	require.NotNil(t, cerr, "Skipchain accepted our fake block!")
+
+	// Unhack: verify that the correct way of doing it works, even if
+	// we bypass the identity.
+	sig, err = crypto.SignSchnorr(network.Suite, c1.Private, hash)
+	log.ErrFatal(err)
+	data2.Votes["one1"] = &sig
+	_, cerr = s0.skipchain.StoreSkipBlock(id.SCData, nil, data2)
+	log.ErrFatal(err)
+	log.ErrFatal(c1.DataUpdate())
+
+	if len(c1.Data.Device) != 2 {
+		t.Fatal("Should have two owners now")
+	}
```

Comment:
If not too much, just adding a test where there is a wrong signature, like random bytes, would be nice :)